lib/winpki/ldap.go

```diff
@@ -174,35 +182,159 @@ func convertLDAPError(err error) error {
+		}
+	}
+	if len(entries) == 0 {
+		return "", "", trace.NotFound("could not find Windows account %q", username)
+	}
+	if len(entries) > 1 {
+		l.cfg.Logger.WarnContext(ctx, "found multiple entries for user, taking the first", "username", username)
+	}
+	activeDirectorySID, err := ADSIDStringFromLDAPEntry(entries[0])
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	l.cfg.Logger.DebugContext(ctx, "Found objectSid Windows user", "username", username)
+	distinguishedName := entries[0].DN
+	return activeDirectorySID, distinguishedName, nil
+}
+
+func (l *LDAPClient) queryLDAP(ctx context.Context, filter string, username string, domainDN string, followReferrals bool) ([]*ldap.Entry, error) {
+	filter = CombineLDAPFilters([]string{
 		fmt.Sprintf("(%s=%s)", AttrSAMAccountType, AccountTypeUser),
-		fmt.Sprintf("(%s=%s)", AttrSAMAccountName, ldap.EscapeFilter(username)),
+		filter,
 	})
+	l.cfg.Logger.DebugContext(ctx, "querying LDAP for objectSid of Windows user", "username", username, "filter", filter, "domain", domainDN)
+
+	entries, err := l.ReadWithFilter(ctx, domainDN, filter, []string{AttrObjectSid}, followReferrals)
+
+	return entries, err
+}
 
-	entries, err := l.ReadWithFilter(DomainDN(l.cfg.Domain), filter, []string{AttrObjectSid})
-	switch {
-	case err != nil:
-		return "", trace.Wrap(err)
-	case len(entries) == 0:
-		return "", trace.NotFound("could not find Windows account %q", username)
-	case len(entries) > 1:
-		l.cfg.Logger.WarnContext(ctx, "found multiple entries for user, taking the first", "user", username)
+// extractReferrals gathers referrals from ldapErr
+// If LDAP server can't provide the information required but has the knowledge of proper it will return error like:
+// LDAP Result Code 10 "Referral": 0000202B: RefErr: DSID-0310084A
+// You then have to parse content of the ber-encoded error to extract the address for the referral
+func extractReferrals(ldapErr *ldap.Error) []string {
+	if ldapErr == nil {
+		return nil
 	}
 
-	sid, err := ADSIDStringFromLDAPEntry(entries[0])
-	if err != nil {
-		return "", trace.Wrap(err)
+	if ldapErr.ResultCode != ldap.LDAPResultReferral {
+		return nil
+	}
+	searchResultIndex := slices.IndexFunc(ldapErr.Packet.Children, func(packet *ber.Packet) bool {
+		return packet.Description == "Search Result Done"
+	})
+	if searchResultIndex < 0 {
+		return nil
 	}
+	searchResult := ldapErr.Packet.Children[searchResultIndex]
 
-	return sid, nil
+	referralsIndex := slices.IndexFunc(searchResult.Children, func(packet *ber.Packet) bool {
+		return packet.Description == "Referral"
+	})
+	if referralsIndex < 0 {
+		return nil
+	}
+	referrals := searchResult.Children[referralsIndex].Children
+
+	out := make([]string, 0, len(referrals))
+	for _, referral := range referrals {
+		referralValue, ok := referral.Value.(string)
+		// we only support LDAPS connections
+		if ok && strings.HasPrefix(referralValue, "ldaps://") {
+			out = append(out, referralValue)
+		}
+	}
+
+	return out
```

Comment:
Should we de-duplicate these just to be safe? I could see some broken setup somewhere creating a cycle of referrals and sending us on an infinite (or excessively large) loop.